### PR TITLE
bug-fix: delay flushing undo files until after they are finalized

### LINKED ISSUE #17890: Non-latest rev files are sometimes re-opened but never re-flushed
The code currently assumes that once a block file is 'left', it and its corresponding rev file can be flushed, and won't be written to any further.

https://github.com/bitcoin/bitcoin/blob/e8e79958a7b2a0bf1b02adcce9f4d811eac37dfc/src/validation.cpp#L1742-L1743

(Flushing here additionally means to get rid of pre-allocated unused space.)

This is the case for the bl(oc)k files, but it turns out to not be the case for the rev files. During an IBD up to ~200k blocks, the following excerpt (one of several) was encountered (grepping out irrelevancy):
```
[...]
2020-01-07T14:19:09Z Pre-allocating up to position 0x7000000 in blk00013.dat
2020-01-07T14:19:13Z Pre-allocating up to position 0x8000000 in blk00013.dat
2020-01-07T14:19:14Z Pre-allocating up to position 0xf00000 in rev00013.dat
2020-01-07T14:19:16Z Leaving block file 13: CBlockFileInfo(blocks=1099, size=134068483, heights=183531...185655, time=2012-06-08...2012-06-22)
2020-01-07T14:19:16Z Pre-allocating up to position 0x1000000 in blk00014.dat
2020-01-07T14:19:17Z Pre-allocating up to position 0x100000 in rev00014.dat
2020-01-07T14:19:20Z Pre-allocating up to position 0x1000000 in rev00013.dat
2020-01-07T14:19:20Z Pre-allocating up to position 0x2000000 in blk00014.dat
2020-01-07T14:19:24Z Pre-allocating up to position 0x1100000 in rev00013.dat
2020-01-07T14:19:26Z Pre-allocating up to position 0x3000000 in blk00014.dat
2020-01-07T14:19:27Z Pre-allocating up to position 0x1200000 in rev00013.dat
2020-01-07T14:19:30Z Pre-allocating up to position 0x200000 in rev00014.dat
2020-01-07T14:19:33Z Pre-allocating up to position 0x300000 in rev00014.dat
2020-01-07T14:19:35Z Pre-allocating up to position 0x400000 in rev00014.dat
2020-01-07T14:19:37Z Pre-allocating up to position 0x500000 in rev00014.dat
2020-01-07T14:19:39Z Pre-allocating up to position 0x600000 in rev00014.dat
2020-01-07T14:19:42Z Pre-allocating up to position 0x700000 in rev00014.dat
2020-01-07T14:19:43Z Pre-allocating up to position 0x4000000 in blk00014.dat
2020-01-07T14:19:45Z Pre-allocating up to position 0x800000 in rev00014.dat
2020-01-07T14:19:49Z Pre-allocating up to position 0x900000 in rev00014.dat
[...]
```

But the only time we flush (and truncate the files to their correct size) is when we leave a block, and we already left block 13 here, so the rev files above end up with pre-allocated space that is never used. This is related to #17827.

This seems to repeat itself for every file that ended up larger than expected in https://github.com/bitcoin/bitcoin/pull/17887#issuecomment-571591432. (Note: #17887 is not a fix for this issue. #17887 makes the impact much smaller by fixing a separate blow-up bug, but it does not address re-flushing re-opened rev files.)

### PR DESCRIPTION
**Pull request by kallewoof**
The blk files have blocks written to them as they come in. Each blk file and rev file reference the same blocks, but the order in which block data is written into the rev (undo) files is in order of activation, i.e. in block height order.

Thus, when the code decides to stop writing into blk000X.dat, it will flush this and rev000X.dat, and then move to X+1. The problem is that since blocks are activated later than they are stored in the block file, we will still be writing to rev X for awhile after this.

This code fixes this by simply finalizing the *previous* undo file each time we finish up a block file. This pretty much guarantees that we will never pre-allocate space and never flush for an undo file (which is happening right now).

Unfortunately for anyone with affected machines, there will be wasted drive space for each rev file if they did IBD before this fix. Probably affects all architectures.

Edit: note that this PR initially added a new critical section for undo pos and some other stuff, but I scratched that for this simpler solution, hence practicalswift's comment below.

Fixes #17890.

### CODE REVIEW

**Comment by practicalswift** [comment]
Thread-safety annotations (`GUARDED_BY` etc.) seems to be missing? :)

**Comment by kallewoof** [comment]
@practicalswift Thanks for the nudge. My apologies, though... I ended up scratching this solution for a much simpler one (see updated code).

**Comment by kallewoof** [comment]
Some stats: after syncing my mac up to ~ block 186k, the blocks dir looks like this:
```
du -ch *.dat
128M	blk00000.dat
128M	blk00001.dat
128M	blk00002.dat
128M	blk00003.dat
128M	blk00004.dat
128M	blk00005.dat
128M	blk00006.dat
128M	blk00007.dat
128M	blk00008.dat
128M	blk00009.dat
128M	blk00010.dat
128M	blk00011.dat
128M	blk00012.dat
128M	blk00013.dat
144M	blk00014.dat
 19M	rev00000.dat
 16M	rev00001.dat
 16M	rev00002.dat
 16M	rev00003.dat
 16M	rev00004.dat
 17M	rev00005.dat
 17M	rev00006.dat
 17M	rev00007.dat
 18M	rev00008.dat
 17M	rev00009.dat
 17M	rev00010.dat
 17M	rev00011.dat
 18M	rev00012.dat
171M	rev00013.dat
 21M	rev00014.dat
2.3G	total
```
The fact the blk for 14 and rev files for 13-14 are bloated is due to a separate bug (see #17887).

Note that rev 13 is still huge despite us moving on to rev 14. When blk 14 is *finalized*, it will truncate rev 13 down to its right size. It may be worthwhile to actually finalize current and previous twice, to avoid this, but since we are more or less guaranteed to again pre-allocate into X-1, this seems unnecessary, and #17887 will make this have significantly less impact.

**Comment by maaku** [comment]
I can confirm the bug. I've run into it multiple times. Kalle is convinced that we are at fault, although I'm not so sure, and I think this might even be an Apple bug. It only shows up on APFS, and the implementation of rsync which ships with macOS also suffers from the same issue.

But in any case, if we can fix it on our end we should, and this should be pretty high priority. It can end up wasting hundreds of gigabytes of space during an IBD, in some cases exhausting the file on 512GB drives in a way that is very non-intuitive to figure out and fix as a user.

**Comment by eriknylund** [comment]
ACK 0508072e1a45c339a181d24c2edfc88e3cd5201f built locally. All tests pass. I see the blow-up happening and the files being truncated as the following blk file is opened and the previous is finalized.

Here rev00004.dat is still not truncated since we are on blk00005.dat
```
128M	blk00000.dat
128M	blk00001.dat
128M	blk00002.dat
128M	blk00003.dat
128M	blk00004.dat
576M	blk00005.dat
 19M	rev00000.dat
 16M	rev00001.dat
 16M	rev00002.dat
 16M	rev00003.dat
153M	rev00004.dat
136M	rev00005.dat
```

In the next step blk00005.dat has been finalized and rev00004.dat is truncated.
```
128M	blk00000.dat
128M	blk00001.dat
128M	blk00002.dat
128M	blk00003.dat
128M	blk00004.dat
128M	blk00005.dat
 16M	blk00006.dat
 19M	rev00000.dat
 16M	rev00001.dat
 16M	rev00002.dat
 16M	rev00003.dat
 16M	rev00004.dat
153M	rev00005.dat
1.0M	rev00006.dat
```

File: src/validation.cpp [hunk 1]
```diff
@@ -1735,8 +1735,14 @@ void static FlushBlockFile(bool fFinalize = false)
 {
     LOCK(cs_LastBlockFile);
 
+    // we want to finalize flush the previous undo file, as it will have been written to
+    // for awhile even after we've moved onto the next blk file; if we don't do this, we
+    // end up with a re-opened but unflushed rev file, resulting in pre-allocated but not
+    // returned disk space allocation
+    int last_undo_file = fFinalize && nLastBlockFile > 0 ? nLastBlockFile - 1 : nLastBlockFile;
```
Comment (eriknylund) [inline src/validation.cpp]:
```suggestion
    int previous_undo_file = fFinalize && nLastBlockFile > 0 ? nLastBlockFile - 1 : nLastBlockFile;
```
Given the comment on line 1738, is _previous_ more appropriate than last?

**Review by eriknylund — CHANGES_REQUESTED** [review]

File: src/validation.cpp (hunk 1 above)
Comment (kallewoof) [reply src/validation.cpp]: Makes sense.

**Comment by fjahr** [comment]
Similar to here https://github.com/bitcoin/bitcoin/pull/17887#issuecomment-576780731 I am not able to reproduce this issue on current master (no matter if `--enable-debug` or not). Happy to take hints and retest as I am curious why my results are different from everyone else.

**Comment by kallewoof** [comment]
@fjahr  https://github.com/bitcoin/bitcoin/pull/17887#issuecomment-576780731 seems to indicate you were finally able to reproduce the issue. Or is that comment older?

**Comment by fjahr** [comment]
> @fjahr [#17887 (comment)](https://github.com/bitcoin/bitcoin/pull/17887#issuecomment-576780731) seems to indicate you were finally able to reproduce the issue. Or is that comment older?

Yes, sorry, this is not worded clearly: I could reproduce #17887 with rev files up 64M large if compiled without `--enable-debug` but I could not reproduce it with `--enable-debug`. This one I can not reproduce no matter if I set debugging or not.

**Comment by kallewoof** [comment]
@fjahr You can not detect the following pattern anywhere in your log file even after syncing for awhile?

```
(date) Leaving block file X: CBlockFileInfo(...)
(date) Pre-allocating up to position 0x1000000 in blk(X+1).dat
(date) Pre-allocating up to position 0x100000 in rev(X+1).dat
[...]
(date) Pre-allocating up to position (...) in rev(X).dat
```

**Comment by fjahr** [comment]
@kallewoof I do see it, but not regularly. From my last sync up to 190k it seems to appear 3 times: For X=7, X=10 and X=14. I was trying hard to catch a larger than normal file size with `du -ch *.dat`, so I missed the first block/rev file transition?

**Comment by kallewoof** [comment]
There's some amount of randomness involved, as the order in which blocks are received and the order in which they are connected is different.

With #17887 merged, it will be very hard to physically spot these now, but it should occasionally show larger-than-necessary rev files (in the range of 1 b up to 1 MB).

**Comment by Sjors** [comment]
Given that #17887 is it still worth the risk of back-porting?

File: src/validation.cpp [hunk 2]
```diff
@@ -1735,8 +1735,14 @@ void static FlushBlockFile(bool fFinalize = false)
 {
     LOCK(cs_LastBlockFile);
 
+    // we want to finalize flush the previous undo file, as it will have been written to
+    // for awhile even after we've moved onto the next blk file; if we don't do this, we
+    // end up with a re-opened but unflushed rev file, resulting in pre-allocated but not
+    // returned disk space allocation
+    int previous_undo_file = fFinalize && nLastBlockFile > 0 ? nLastBlockFile - 1 : nLastBlockFile;
```
Comment (Sjors) [inline src/validation.cpp]: nit: add brackets to clarify (correct) precedence: `(nLastBlockFile > 0)`

**Comment by kallewoof** [comment]
@Sjors

> Given that #17887 is merged, is it still worth the risk of back-porting? (can always do that later, after 0.20 is released)

There are nearly two thousand rev files right now. Someone doing an IBD sync using 0.19 today would end up wasting up to 2 GB of space due to this bug. I do believe we want to plug that hole, even though it has become much smaller of an issue as you mention.

File: src/validation.cpp (hunk 2 above)
Comment (kallewoof) [reply src/validation.cpp]: I honestly don't think that's necessary, but I'm open to change my opinion.

**Comment by maaku** [comment]
Last time I did an IBD (a few months ago) it was >100GB wasted. That’s actually how I discovered this problem initially, it was enough wasted space to impact me (full disk).

**Comment by kallewoof** [comment]
@maaku 

> Last time I did an IBD (a few months ago) it was >100GB wasted. That’s actually how I discovered this problem initially, it was enough wasted space to impact me (full disk).

Yeah, that was partially fixed in #17887, so you shouldn't see quite that drastic of an effect anymore, but still..

**Comment by maaku** [comment]
And this discussion is about backporting those changes, no?

**Comment by kallewoof** [comment]
No, #17887 is already in the backported list and is not under discussion. @Sjors is wondering whether this patch, which delays finalize-flushing the rev files until we're (pretty) sure they are no longer appended to, should be backported or not.

Without this patch, we will experience 0-1 MB of wasted space for each rev file.

Without #17887 we would experience what you saw.

**Comment by fjahr** [comment]
I have now done another run with this fix and could not detect the pattern described in https://github.com/bitcoin/bitcoin/pull/17892#issuecomment-577243156 in my debug.log

ACK b49b34139ac3c187eaac21b25f88cd493ce668b2

**Comment by MarcoFalke** [comment]
> Without this patch, we will experience 0-1 MB of wasted space for each rev file.

It seems like this does not warrant a backport to 0.19.1

**Comment by kallewoof** [comment]
@MarcoFalke 

Depends if you think "a few GBs wasted" is negligible or not.

**Comment by MarcoFalke** [comment]
Yes, based on your comments this should be less than 1% off (not noticeable for a user)

**Comment by anton48** [comment]
> @MarcoFalke
> 
> Depends if you think "a few GBs wasted" is negligible or not. (This is for non-pruned nodes only, so a few GBs in a 250 GB datadir.)

I did a quick test with downloading from scratch up to year 2004 (approx. 270k blocks):

**master (278292 blocks, 1931064 bytes, 10 files with size 19M)**

```
2020-01-23T11:01:43Z Leaving block file 104: CBlockFileInfo(blocks=871, size=134096120, heights=276943...278292, time=2013-12-25...2014-01-02)

Antons-Mac-mini:HDD2 a48$ ls -la Bitcoin.nopatch/blocks/rev00* | wc -l
     105

Antons-Mac-mini:HDD2 a48$ du -ck Bitcoin.nopatch/blocks/rev00* | grep total
1931064	total

Antons-Mac-mini:HDD2 a48$ du -ch Bitcoin.nopatch/blocks/rev00* | grep 19M
 19M	Bitcoin.nopatch/blocks/rev00000.dat
 19M	Bitcoin.nopatch/blocks/rev00013.dat
 19M	Bitcoin.nopatch/blocks/rev00073.dat
 19M	Bitcoin.nopatch/blocks/rev00075.dat
 19M	Bitcoin.nopatch/blocks/rev00081.dat
 19M	Bitcoin.nopatch/blocks/rev00083.dat
 19M	Bitcoin.nopatch/blocks/rev00086.dat
 19M	Bitcoin.nopatch/blocks/rev00087.dat
 19M	Bitcoin.nopatch/blocks/rev00094.dat
 19M	Bitcoin.nopatch/blocks/rev00101.dat
Antons-Mac-mini:HDD2 a48$ du -ch Bitcoin.nopatch/blocks/rev00* | grep 19M | wc -l
      10
Antons-Mac-mini:HDD2 a48$ du -ch Bitcoin.nopatch/blocks/rev00* | grep 18M | wc -l
      80
```

**this PR (278293 blocks, 1865004 bytes, no files with size 19M, all of them are 17M or 18M)**

```
2020-01-23T12:48:45Z Leaving block file 104: CBlockFileInfo(blocks=865, size=134045380, heights=277411...278293, time=2013-12-28...2014-01-02)

Antons-Mac-mini:HDD2 a48$ ls -la Bitcoin.patch/blocks/rev00* | wc -l
     105

Antons-Mac-mini:HDD2 a48$ du -ck Bitcoin.patch/blocks/rev00* | grep total
1865004	total

Antons-Mac-mini:HDD2 a48$ du -ch Bitcoin.patch/blocks/rev00* | grep 19M
Antons-Mac-mini:HDD2 a48$ du -ch Bitcoin.patch/blocks/rev00* | grep 19M | wc -l
       0
Antons-Mac-mini:HDD2 a48$ du -ch Bitcoin.patch/blocks/rev00* | grep 18M | wc -l
      38
```

so, it looks for me that there could be more than 1G of overhead on a fullnode with all blocks (will test it, but it takes time).

**Comment by kallewoof** [comment]
@anton48 Thanks for the initiative. I'll do the same thing on my side so we can get some clarity on the matter.

@MarcoFalke I agree with you, but with the caveat that it's temporary. It's ludicrous to waste several GB of space on people's machines for no good reason. My proposal is that we (1) make a PR that does a one time sweep (esp important for people with lots of disk space who are stuck with pre-#17887 datadirs, if such people exist at all) + implements this code*, and to potentially backport those two.

(* or possibly more sophisticated; ideal implementation is to, for each blk file, track which block is the highest height, and once that has been connected, to finalize the corresponding rev file)

**Comment by Sjors** [comment]
> I agree with you, but with the caveat that it's temporary. It's ludicrous to waste several GB of space on people's machines for no good reason. 

The waste would be temporary; until the user install 0.20 or 0.19.2. Worst case scenario is block data corruption.Because I'm personally not familiar with this area of the code, I prefer to let it stew on master for a while before shipping. But if others feel very comfortable that this code is correct then I'm fine with backporting.

**Comment by eriknylund** [comment]
> @anton48 Thanks for the initiative. I'll do the same thing on my side so we can get some clarity on the matter.
> 
> @MarcoFalke I agree with you, but with the caveat that it's temporary. It's ludicrous to waste several GB of space on people's machines for no good reason. My proposal is that we (1) make a PR that does a one time sweep (esp important for people with lots of disk space who are stuck with pre-#17887 datadirs, if such people exist at all) + implements this code*, and to potentially backport those two.
> 
> (* or possibly more sophisticated; ideal implementation is to, for each blk file, track which block is the highest height, and once that has been connected, to finalize the corresponding rev file)

Given it has not been getting attention since the release of APFS in September, 2017 it seems not that many installations exist. Is it worth adding this extra complexity for little or no users? Do we know how  many are potentially running a version that has the bug? The workarounds are not optimal either of course, but at least there are two relatively easy ones:
1. Do the IBD from scratch using the lastest version
2. Run a truncate bash script (https://github.com/bitcoin/bitcoin/issues/17827#issuecomment-570125891)

**Comment by eriknylund** [comment]
@kallewoof @anton48 I can also repeat the full IBD test running Bitcoin-Qt on the Mac Mini like I did for #17887, it will take about 3 days to complete.

**Comment by kallewoof** [comment]
@eriknylund Unfortunately that bash script only addresses #17887, not #17892.

@Sjors It's not technically temporary until someone writes code that does a one-time sweep of people's rev files and backport this & that, but since we only see ~1 rev file a day, people's drives shouldn't suffer more than a few hundred MBs before a patch is in place, hopefully.

**Comment by kallewoof** [comment]
I am working on an improved version of this, based on https://github.com/bitcoin/bitcoin/pull/17892#issuecomment-577703536

**Comment by eriknylund** [comment]
> @eriknylund Unfortunately that bash script only addresses #17887, not #17892.
> 
> @Sjors It's not technically temporary until someone writes code that does a one-time sweep of people's rev files and backport this & that, but since we only see ~1 rev file a day, people's drives shouldn't suffer more than a few hundred MBs before a patch is in place, hopefully.

@kallewoof Ah you're absolutely correct. Thanks for pointing that out.

**Comment by anton48** [comment]
> > @eriknylund Unfortunately that bash script only addresses #17887, not #17892.
> > @Sjors It's not technically temporary until someone writes code that does a one-time sweep of people's rev files and backport this & that, but since we only see ~1 rev file a day, people's drives shouldn't suffer more than a few hundred MBs before a patch is in place, hopefully.
> 
> @kallewoof Ah you're absolutely correct. Thanks for pointing that out.

however, in my case the script made some truncations at yesterday test directory (which is on AFPS volume):

```
Antons-Mac-mini:blocks a48$ /Users/a48/slim.sh
./rev00000.dat 19M  ->  19M
[..]
./rev00013.dat 19M  ->  18M
[..]
./rev00047.dat 18M  ->  17M
[..]
./rev00053.dat 18M  ->  17M
[..]
./rev00063.dat 18M  ->  17M
./rev00064.dat 18M  ->  17M
[..]
./rev00070.dat 18M  ->  17M
[..]
./rev00073.dat 19M  ->  18M
[..]
./rev00075.dat 19M  ->  18M
[..]
./rev00083.dat 19M  ->  18M
[..]
./rev00086.dat 19M  ->  18M
./rev00087.dat 19M  ->  18M
[..]
./rev00101.dat 19M  ->  18M
[..]
./rev00104.dat 18M  ->  18M
```

**Comment by kallewoof** [comment]
I did some number crunching and concluded that a backport is probably not necessary. I am also closing this PR in favor of #17994. Will re-open if people object/prefer this one.

**Pull request closed without merging**